execute/report/report.go

```diff
@@ -118,81 +121,144 @@ func buildSingleChainReport(
+		InvalidNonce                         messageStatus = "invalid_nonce"
+		AggregateTokenValueComputeError      messageStatus = "aggregate_token_value_compute_error"
+		AggregateTokenLimitExceeded          messageStatus = "aggregate_token_limit_exceeded"
+		TokenDataNotReady                    messageStatus = "token_data_not_ready"
+		TokenDataFetchError                  messageStatus = "token_data_fetch_error"
+		TokenNotInDestTokenPrices            messageStatus = "token_not_in_dest_token_prices"
+		TokenNotInSrcTokenPrices             messageStatus = "token_not_in_src_token_prices"
+		InsufficientRemainingFee             messageStatus = "insufficient_remaining_fee"
+		AddedToBatch                         messageStatus = "added_to_batch"
+	*/
+)
+
+func (b *execReportBuilder) checkMessage(
+	_ context.Context, idx int, execReport plugintypes.ExecutePluginCommitDataWithMessages,
+	// TODO: get rid of the nolint when the error is used
+) (messageStatus, error) { // nolint this will use the error eventually
+	if slices.Contains(execReport.ExecutedMessages, execReport.Messages[idx].Header.SequenceNumber) {
+		return AlreadyExecuted, nil
+	}
+
+	return ReadyToExecute, nil
+}
+
+func (b *execReportBuilder) verifyReport(
+	ctx context.Context, execReport cciptypes.ExecutePluginReportSingleChain,
+) (bool, validationMetadata, error) {
 	// Compute the size of the encoded report.
-	encoded, err := encoder.Encode(
+	// Note: ExecutePluginReport is a strict array of data, so wrapping the final report
+	//       does not add any additional overhead to the size being computed here.
+	encoded, err := b.encoder.Encode(
 		ctx,
 		cciptypes.ExecutePluginReport{
-			ChainReports: []cciptypes.ExecutePluginReportSingleChain{finalReport},
+			ChainReports: []cciptypes.ExecutePluginReportSingleChain{execReport},
 		},
 	)
 	if err != nil {
-		lggr.Errorw("unable to encode report", "err", err, "report", finalReport)
-		return cciptypes.ExecutePluginReportSingleChain{}, 0, fmt.Errorf("unable to encode report: %w", err)
+		b.lggr.Errorw("unable to encode report", "err", err, "report", execReport)
+		return false, validationMetadata{}, fmt.Errorf("unable to encode report: %w", err)
+	}
+
+	maxSizeBytes := int(b.maxReportSizeBytes - b.accumulated.encodedSizeBytes)
+	if len(encoded) > maxSizeBytes {
+		b.lggr.Debugw("invalid report, report size exceeds limit", "size", len(encoded), "maxSize", maxSizeBytes)
+		return false, validationMetadata{}, nil
 	}
 
-	return finalReport, len(encoded), nil
+	return true, validationMetadata{
+		encodedSizeBytes: uint64(len(encoded)),
+	}, nil
 }
 
-// buildSingleChainReportMaxSize generates the largest report which fits into maxSizeBytes.
+// buildSingleChainReport generates the largest report which fits into maxSizeBytes.
 // See buildSingleChainReport for more details about how a report is built.
-func buildSingleChainReportMaxSize(
+func (b *execReportBuilder) buildSingleChainReport(
```

Comment:
Changed to be a receiver method, so most of the arguments go away. Some functionality has been moved to the new `checkMessage` and `verifyReport` methods.

Reply:
nit/ Both the internal and the interface `buildSingleChainReport` methods are named the same.

What do you think of `buildSingleChainReportHelper`?

Reply:
Good idea, renamed.